File: 2024/10_Hoof_It/Day10.js
// Day 10: Hoof It
// https://adventofcode.com/2024/day/10

const fs = require('fs');
var readInput = function (filename) {
  return fs.readFileSync(`${filename}.txt`, 'utf8').split('\n').filter(Boolean);
};

function buildGrid(input) {
  return input.map((row) => row.trim().split('').map(Number));
}

function countNinthPoints(grid, countUniquePoint) {
  const rows = grid.length;
  const cols = grid[0].length;
  const directions = [0, 1, 0, -1, 0];

  let totalScore = 0;
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (grid[row][col] !== 0) continue;

      const ninthPoints = getNinthPoints(row, col);
      if (countUniquePoint) {
        const uniqueNinthPoints = new Set(ninthPoints);
        totalScore += uniqueNinthPoints.size;
      } else {
        totalScore += ninthPoints.length;
      }
    }
  }
  return totalScore;

  function isInBounds(row, col) {
    return row >= 0 && row < rows && col >= 0 && col < cols;
  }

  function getNinthPoints(startRow, startCol) {
    const ninthPoints = [];
    let queue = [[startRow, startCol, 0]];
    while (queue.length > 0) {
      const nextQueue = [];
      for (const [row, col, currentNumber] of queue) {
        if (currentNumber === 9) {
          const key = row * cols + col;
          ninthPoints.push(key);
          continue;
        }

        for (let d = 0; d < 4; d++) {
          const newRow = row + directions[d];
          const newCol = col + directions[d + 1];
          const nextNumber = currentNumber + 1;

          if (!isInBounds(newRow, newCol)) continue;
          if (grid[newRow][newCol] !== nextNumber) continue;

          nextQueue.push([newRow, newCol, nextNumber]);
        }
      }
      queue = nextQueue;
    }
    return ninthPoints;
  }
}

function part1(input) {
  const grid = buildGrid(input);
  return countNinthPoints(grid, true);
}

function part2(input) {
  const grid = buildGrid(input);
  return countNinthPoints(grid, false);
}

var filename = 'Day10';
console.log('Part 1:', part1(readInput(filename))); // 36
console.time('Part 1');
console.log('Part 1 Test:', part1(readInput(`${filename}_test`))); // 825
console.timeEnd('Part 1');

console.log('Part 2:', part2(readInput(filename))); // 81
console.time('Part 2');
console.log('Part 2 Test:', part2(readInput(`${filename}_test`))); // 1805
console.timeEnd('Part 2');

File: 2024/10_Hoof_It/Day10.kt
// Day 10: Hoof It
// https://adventofcode.com/2024/day/10

import java.io.File

fun readInput(filename: String) = File("$filename.txt").readText().trim()

fun buildGrid(input: String) = input.lines().map { it.trim().map { it.digitToInt() } }

fun getNinthPointsFromZeroPoints(grid: List<List<Int>>) = grid.flatMapIndexed { row, line ->
    line.mapIndexedNotNull { col, value ->
        if (value == 0) getNinthPoints(row, col, grid) else null
    }
}

fun getNinthPoints(startRow: Int, startCol: Int, grid: List<List<Int>>): List<Int> {
    val rows = grid.size
    val cols = grid[0].size
    val directions = listOf(0, 1, 0, -1, 0)
    val ninthPoints = mutableListOf<Int>()
    val queue = ArrayDeque<Triple<Int, Int, Int>>()
    queue.add(Triple(startRow, startCol, 0))

    fun isInBounds(row: Int, col: Int) = row in 0 until rows && col in 0 until cols

    while (queue.isNotEmpty()) {
        val (row, col, currentNumber) = queue.removeFirst()

        if (currentNumber == 9) {
            val key = row * cols + col
            ninthPoints.add(key)
            continue
        }

        for (d in 0 until 4) {
            val newRow = row + directions[d]
            val newCol = col + directions[d + 1]
            val nextNumber = currentNumber + 1

            if (isInBounds(newRow, newCol) && grid[newRow][newCol] == nextNumber) {
                queue.add(Triple(newRow, newCol, nextNumber))
            }
        }
    }
    return ninthPoints
}

fun part1(input: String): Int {    
    val grid = buildGrid(input)
    val ninthPoints = getNinthPointsFromZeroPoints(grid)
    return ninthPoints.map { it.toSet().size }.sum()
}

fun part2(input: String): Int {
    val grid = buildGrid(input)
    val ninthPoints = getNinthPointsFromZeroPoints(grid)
    return ninthPoints.flatten().size
}

fun main() {
    val input = readInput("Day10")
    val inputTest = readInput("Day10_test")

    println("Part 1: ${part1(input)}") // 36
    println("Part 1 Test: ${part1(inputTest)}") // 825

    println("Part 2: ${part2(input)}") // 81
    println("Part 2 Test: ${part2(inputTest)}") // 1805
}
